packages/ui/src/components/va-tree-view/hooks/useTreeKeyboardNavigation.ts

```diff
@@ -0,0 +1,180 @@
+import { ExtractPropTypes } from 'vue'
+
+import type { TreeNode, TreeView, TreeViewEmitsFunc } from '../types'
+import { useTreeViewProps } from './useTreeHelpers'
+
+type TreeNodeElement = HTMLElement | null
+type UseTreeKeyboardNavigationMethods = Pick<TreeView, 'toggleNode' | 'toggleCheckbox'> & {
+  emit: TreeViewEmitsFunc,
+}
+
+const useTreeKeyboardNavigation = (
+  props: ExtractPropTypes<typeof useTreeViewProps>,
+  methods: UseTreeKeyboardNavigationMethods,
+) => {
+  const { emit, toggleNode, toggleCheckbox } = methods
+  const isElementExpanded = (currentElement: TreeNodeElement): boolean =>
+    currentElement?.getAttribute('aria-expanded') === 'true' || false
+
+  const getParentElement = (currentElement: TreeNodeElement): TreeNodeElement =>
+    currentElement?.parentElement?.closest('.va-tree-node') || null
+
+  const getPreviousElement = (currentElement: TreeNodeElement): TreeNodeElement =>
+    currentElement?.previousElementSibling as TreeNodeElement
+
+  const findPreviousElement = (currentElement: TreeNodeElement): TreeNodeElement => {
+    if (!currentElement) {
+      return null
+    }
+
+    let previousElement = getPreviousElement(currentElement)
+    let lastChildElement = isElementExpanded(previousElement) && getLastChildElement(previousElement)
+
+    if (lastChildElement) {
+      do {
+        if (isElementExpanded(lastChildElement)) {
```

Comment:
Imo it's very deep - can remove 1-2 levels via early break.

Reply:
I recommend avoiding while loop in general and limit usage of for loop to the minimum. Everything that can be done by those can by done in more readable declarative way using array methods.

Reply:
Though I'm not sure there are other ways if it's linked list type of thing you're dealing with.

Reply:
Yep, it's kinda a linked list.